### PR TITLE
make accessToken an io to use gh app's expiring tokens

### PR DESCRIPTION
**Pull request by reimai**
Hi. In my project I want to use Github App auth, which operates expiring tokens. So here is a proposed change to the insides of the api:
* move accessToken into the http client 
* make it a monad

The default Github constructor stays the same, it's accessToken is just wrapped into a pure version of AccessTokens.
I've read you don't want to [bring things like jwt on board](https://github.com/47degrees/github4s/issues/554), so the implementation of AccessTokens for apps is left for users.

### CODE REVIEW

File: github4s/src/main/scala/github4s/algebras/AccessTokens.scala [hunk 1]
```diff
@@ -0,0 +1,13 @@
+package github4s.algebras
+
+import github4s.GHResponse
+
+/**
+ * Source of static or expiring github tokens
+ */
+trait AccessTokens[F[_]] {
```
Comment (BenFradet) [inline github4s/src/main/scala/github4s/algebras/AccessTokens.scala:8]:
```suggestion
trait AccessToken[F[_]] {
```

File: github4s/src/main/scala/github4s/algebras/AccessTokens.scala [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+
+  def withAccessToken[T](f: Option[String] => F[GHResponse[T]]): F[GHResponse[T]]
+}
+
+object AccessTokens {}
```
Comment (BenFradet) [inline github4s/src/main/scala/github4s/algebras/AccessTokens.scala:13]:
```suggestion
```

File: github4s/src/main/scala/github4s/interpreters/StaticAccessTokens.scala [hunk 3]
```diff
@@ -0,0 +1,17 @@
+package github4s.interpreters
+
+import github4s.GHResponse
+import github4s.algebras.AccessTokens
+
+/**
+ * A simple static version
+ */
+class StaticAccessTokens[F[_]](accessToken: Option[String]) extends AccessTokens[F] {
```
Comment (BenFradet) [inline github4s/src/main/scala/github4s/interpreters/StaticAccessTokens.scala:9]:
```suggestion
class StaticAccessToken[F[_]](accessToken: Option[String]) extends AccessToken[F] {
```

**Review by BenFradet — COMMENTED** [review]
I like the approach :+1: 

Do the integration tests work locally?

Also I feel like we need to adapt the existing docs and create some more for someone looking to extend access token like you.

**Comment by reimai** [comment]
I've renamed AccessTkoens and added some docs, my english might not be perfect though.

I tried running integrations tests locally with my personal token, and some of the tests fail for an obvious reason (me not having enough permissions for the github4s repository). Others are fine.

File: github4s/src/main/scala/github4s/modules/GithubAPIs.scala [hunk 4]
```diff
@@ -45,3 +44,9 @@ class GithubAPIv3[F[_]: Sync](
   override val projects: Projects[F]           = new ProjectsInterpreter[F]
 
 }
+
+object GithubAPIv3 {
+
+  def noAuth[F[_]: Sync](client: Client[F], config: GithubConfig) =
```
Comment (BenFradet) [inline github4s/src/main/scala/github4s/modules/GithubAPIs.scala:50]: missing return type

File: microsite/docs/docs.md [hunk 5]
```diff
@@ -10,6 +10,8 @@ permalink: docs
 
 In order to access the Github API, you will need to have [an access token][access-token] with the
 appropriate scopes (i.e. if you want to create gists, your token will need to have the gist scope).
+A personal token lives forever, github app's one has to be re-issued every hour. 
+Implementation of [github app][github_app] authentication involves 3rd party libraries like jwt and is left for the user.
```
Comment (BenFradet) [inline microsite/docs/docs.md:14]:
```suggestion

A personal token can be valid forever. However, the ones issued for [Github Apps][github-app] have to be re-issued every hour. 
To solve that issue, github4s provides [AccessToken][access-token-scala] which you can hook into to provide authentication for your app using third-party libraries such as JWT implementations, for example.
You can find the default implementation for static tokens in [StaticAccessToken][static-access-token-scala].
```

File: microsite/docs/docs.md [hunk 6]
```diff
@@ -240,6 +242,7 @@ headers which will be added to every request sent to the GitHub API. The user ag
 added by default.
 
 [access-token]: https://github.com/settings/tokens
+[github_app]: https://github.com/settings/apps
```
Comment (BenFradet) [inline microsite/docs/docs.md:245]:
```suggestion
[github-app]: https://github.com/settings/apps
[access-token-scala]: https://github.com/47degrees/github4s/blob/master/github4s/src/main/scala/github4s/algebras/AccessToken.scala
[access-token-scala]: https://github.com/47degrees/github4s/blob/master/github4s/src/main/scala/github4s/interpreters/StaticAccessToken.scala
```

**Review by BenFradet — APPROVED** [review]
Thanks a lot for your great contribution :+1:

**Review by juanpedromoreno — APPROVED** [review]
💯 Thanks for your time @reimai!

**Pull request merged**